src/ILCompiler.ReadyToRun/src/Compiler/ReadyToRunMetadataFieldLayoutAlgorithm.cs

```diff
@@ -4,44 +4,257 @@
+                                    }
+                                    break;
+
+                                default:
+                                    throw new NotImplementedException(field.FieldType.Category.ToString());
+                            }
+
+                            LayoutInt[] layout = (isGcField ? gcStatics : nonGcStatics);
+                            LayoutInt offset = LayoutInt.AlignUp(layout[index], new LayoutInt(alignment));
+                            layout[index] = offset + new LayoutInt(size);
+                            if (fieldsForType == null)
+                            {
+                                fieldsForType = new List<FieldAndOffset>();
+                            }
+                            fieldsForType.Add(new FieldAndOffset(field, offset));
+                        }
+                    }
+                    if (fieldsForType != null)
+                    {
+                        typeToFieldMap.Add(defType, fieldsForType.ToArray());
+                    }
+                }
+
+                LayoutInt blockAlignment = new LayoutInt(MaxPrimitiveFieldSize);
+
+                return new ModuleFieldLayout(
+                    module,
+                    gcStatics: new StaticsBlock() { Size = gcStatics[0], LargestAlignment = blockAlignment },
+                    nonGcStatics: new StaticsBlock() { Size = nonGcStatics[0], LargestAlignment = blockAlignment },
+                    threadGcStatics: new StaticsBlock() { Size = gcStatics[1], LargestAlignment = blockAlignment },
+                    threadNonGcStatics: new StaticsBlock() { Size = nonGcStatics[1], LargestAlignment = blockAlignment },
+                    typeToFieldMap: typeToFieldMap);
+
+                throw new NotImplementedException();
+            }
+
+            protected override int GetKeyHashCode(EcmaModule key)
+            {
+                return key.GetHashCode();
+            }
+
+            protected override int GetValueHashCode(ModuleFieldLayout value)
+            {
+                return value.Module.GetHashCode();
+            }
         }
 
-        protected override void FinalizeRuntimeSpecificStaticFieldLayout(TypeSystemContext context, ref ComputedStaticFieldLayout layout)
+        /// <summary>
+        /// Field layouts for a given EcmaModule.
+        /// </summary>
+        private class ModuleFieldLayout
         {
-            if (layout.NonGcStatics.Size == _initialNonGcStaticsOffset)
+            public EcmaModule Module { get; }
+
+            public StaticsBlock GcStatics { get; }
+
+            public StaticsBlock NonGcStatics { get;  }
+
+            public StaticsBlock ThreadGcStatics { get;  }
+
+            public StaticsBlock ThreadNonGcStatics { get;  }
+
+            public Dictionary<DefType, FieldAndOffset[]> TypeToFieldMap { get; }
```

Comment:
Expose as IReadOnlyDictionary so that it doesn't invite someone thinking it's okay to update this?